azure/scope/machinepool.go

```diff
@@ -153,61 +156,291 @@ func (m *MachinePoolScope) ProvisioningState() infrav1.VMState {
+	}
+
+	existingMachinesByProviderID := make(map[string]infrav1exp.AzureMachinePoolMachine, len(ampml.Items))
+	for _, machine := range ampml.Items {
+		existingMachinesByProviderID[machine.Spec.ProviderID] = machine
+	}
 
-			if nodeStatus.Ready {
-				readyReplicas++
+	// determine which machines need to be created to reflect the current state in Azure
+	azureMachinesByProviderID := m.vmssState.InstancesByProviderID()
+	for key, val := range azureMachinesByProviderID {
+		machine, ok := existingMachinesByProviderID[key]
+		if !ok {
+			m.Info("creating machine", "machine", key)
+			if err := m.createMachine(ctx, val); err != nil {
+				return errors.Wrap(err, "failed creating machine")
+			}
+			continue
+		}
+		machine.Status.LatestModelApplied = val.LatestModelApplied // make sure we have the most recent data
+	}
+
+	deleted := false
+	// delete machines that no longer exist in Azure
+	for key, machine := range existingMachinesByProviderID {
+		machine := machine
+		if _, ok := azureMachinesByProviderID[key]; !ok {
+			deleted = true
+			m.Info("deleting machine", "machine", key)
+			delete(existingMachinesByProviderID, key)
+			if err := m.client.Delete(ctx, &machine); err != nil {
+				return errors.Wrap(err, "failed deleting machine to reduce replica count")
 			}
 		}
 	}
 
-	m.AzureMachinePool.Status.Replicas = readyReplicas
-	m.AzureMachinePool.Spec.ProviderIDList = providerIDs
-	m.AzureMachinePool.Status.Instances = instanceStatuses
+	if deleted {
+		m.Info("deleted machines")
+		// exit early to be less greedy about delete
+		return nil
+	}
+
+	// select machines to delete to lower the replica count
+	toDelete := m.selectMachinesToDelete(existingMachinesByProviderID)
+	for key, machine := range toDelete {
+		machine := machine
+		m.Info("deleting selected machine", "machine", key)
+		if err := m.client.Delete(ctx, &machine); err != nil {
+			return errors.Wrap(err, "failed deleting machine to reduce replica count")
+		}
+	}
+
+	m.Info("done cleaning up machines")
 	return nil
 }
 
-// SaveK8sVersion stores the MachinePool spec K8s version to the AzureMachinePool status
-func (m *MachinePoolScope) SaveK8sVersion() {
-	m.AzureMachinePool.Status.Version = *m.MachinePool.Spec.Template.Spec.Version
+func getProviderIDs(machinesByProviderID map[string]infrav1exp.AzureMachinePoolMachine) []string {
+	ids := make([]string, len(machinesByProviderID))
+	idx := 0
+	for k := range machinesByProviderID {
+		ids[idx] = k
+		idx++
+	}
+	return ids
+}
+
+// selectMachinesToDelete will build a map of machines by provider ID to delete prioritizing in the following order
+// 1) failed machines
+// 2) over-provisioned machines prioritized by out of date models first
+// 3) over-provisioned ready machines
+// 4) ready machines within budget which have out of date models
+func (m *MachinePoolScope) selectMachinesToDelete(machinesByProviderID map[string]infrav1exp.AzureMachinePoolMachine) map[string]infrav1exp.AzureMachinePoolMachine {
```

Comment:
This contains the logic for selecting machines to delete when over-provisioned or upgrading. The `AzureMachinePool` informs the `AzureMachinePoolMachine` to delete and the `AzureMachinePoolMachine` is expected to safely remove itself from the pool.